src/clock_synchronization.c

```diff
@@ -0,0 +1,310 @@
+    validate(ret == LF_OK);
+  }
+}
+
+static void ClockSynchronization_handle_priority_update(ClockSynchronization *self, int src_neighbor, int priority) {
+  LF_DEBUG(CLOCK_SYNC, "Received priority %d from neighbor %d", priority, src_neighbor);
+  int my_old_priority = ClockSynchronization_my_priority(self);
+  self->neighbor_clock[src_neighbor].priority = priority;
+  int my_new_priority = ClockSynchronization_my_priority(self);
+  if (my_new_priority < my_old_priority) {
+    LF_DEBUG(CLOCK_SYNC, "Received priority %d+1=%d less than current priority %d, updating", priority, my_new_priority,
+             my_old_priority);
+    self->master_neighbor_index = src_neighbor;
+    ClockSynchronization_handle_priority_request(self, -1);
+  }
+}
+
+static void ClockSynchronization_handle_delay_request(ClockSynchronization *self, SystemEvent *event) {
+  ClockSyncEvent *payload = (ClockSyncEvent *)event->super.payload;
+  lf_ret_t ret;
+  int src_neighbor = payload->neighbor_index;
+  LF_DEBUG(CLOCK_SYNC, "Handling delay request from neighbor %d", src_neighbor);
+  validate(src_neighbor >= 0);
+  validate(src_neighbor < (int)self->num_neighbours);
+  FederatedConnectionBundle *bundle = self->env->net_bundles[src_neighbor];
+  NetworkChannel *chan = bundle->net_channel;
+  bundle->send_msg.which_message = FederateMessage_clock_sync_msg_tag;
+  bundle->send_msg.message.clock_sync_msg.which_message = ClockSyncMessage_delay_response_tag;
+  bundle->send_msg.message.clock_sync_msg.message.sync_response.time = event->super.tag.time;
+  bundle->send_msg.message.clock_sync_msg.message.sync_response.sequence_number =
+      payload->msg.message.delay_request.sequence_number;
+  ret = chan->send_blocking(chan, &bundle->send_msg);
+  if (ret != LF_OK) {
+    LF_WARN(CLOCK_SYNC, "Failed to send DelayResponse to neighbor %zu", src_neighbor);
+  }
+}
+
+static void ClockSynchronization_handle_sync_response(ClockSynchronization *self, SystemEvent *event) {
+  ClockSyncEvent *payload = (ClockSyncEvent *)event->super.payload;
+  lf_ret_t ret;
+  SyncResponse *msg = &payload->msg.message.sync_response;
+  int src_neighbor = payload->neighbor_index;
+  LF_DEBUG(CLOCK_SYNC, "Handling sync response from neighbor %d", src_neighbor);
+
+  if (msg->sequence_number != self->sequence_number) {
+    LF_WARN(CLOCK_SYNC, "Received out-of-order sync response %d from neighbor %d dropping", msg->sequence_number,
+            src_neighbor);
+    return;
+  }
+
+  self->timestamps.t1 = msg->time;
+  self->timestamps.t2 = event->super.tag.time;
+
+  FederatedConnectionBundle *bundle = self->env->net_bundles[src_neighbor];
+  NetworkChannel *chan = bundle->net_channel;
+  bundle->send_msg.which_message = FederateMessage_clock_sync_msg_tag;
+  bundle->send_msg.message.clock_sync_msg.which_message = ClockSyncMessage_delay_request_tag;
+  bundle->send_msg.message.clock_sync_msg.message.delay_request.sequence_number = ++self->sequence_number;
```

Comment:
Should we increment it here? Makes sense to have the same sequence number for entire round?